### ISSUE TITLE
Question about kws_streaming/model/cnn

### ISSUE THREAD

**Issue by ntyoshi** [google-research/google-research#652]
Hi,

I'm trying to duplicate some models on the paper (http://www.isca-speech.org/archive/interspeech_2015/papers/i15_1478.pdf).
I saw the comment below in [kws_streaming/model/cnn.py](https://github.com/google-research/google-research/blob/master/kws_streaming/models/cnn.py):
```
It is based on paper:
  Convolutional Neural Networks for Small-footprint Keyword Spotting
  http://www.isca-speech.org/archive/interspeech_2015/papers/i15_1478.pdf
 ```
So I thought [kws_streaming/model/cnn.py](https://github.com/google-research/google-research/blob/master/kws_streaming/models/cnn.py) had the similar structure as `cnn-one-fpool3` for example.
But the paper describes cnn+max_pool as a typical cnn architecture and it uses the model that has the structure but [kws_streaming/model/cnn.py](https://github.com/google-research/google-research/blob/master/kws_streaming/models/cnn.py) doesn't seem to have any max pooling layer.
Why doesn't [kws_streaming/model/cnn.py](https://github.com/google-research/google-research/blob/master/kws_streaming/models/cnn.py) have it like tf.keras.MaxPool2D?

In addition, `net = stream.Stream(cell=tf.keras.layers.Flatten())(net)` and `net = tf.keras.layers.Dropout(rate=flags.dropout1)(net)` mean  a linear low-rank layer in the paper?

Thank you for your help!

**Comment by rybakov** [comment]
Thank you for evaluating kws_streaming lib!

For cnn model we referred two papers: [Convolutional Neural Networks for Small-footprint Keyword Spotting](http://www.isca-speech.org/archive/interspeech_2015/papers/i15_1478.pdf) and [Hello Edge: Keyword Spotting on Microcontrollers](https://arxiv.org/pdf/1711.07128.pdf), the first one as earliest cnn application and the second one is more similar to what we use in the code. So I would say that cnn model is inspired by [Convolutional Neural Networks for Small-footprint Keyword Spotting](http://www.isca-speech.org/archive/interspeech_2015/papers/i15_1478.pdf) and is based on [Hello Edge: Keyword Spotting on Microcontrollers](https://arxiv.org/pdf/1711.07128.pdf). If you need max pooling, I would suggest to try [ds_tc_resnet](https://github.com/google-research/google-research/blob/master/kws_streaming/models/ds_tc_resnet.py) option, here you can configure you model with and without max pooling.

net = stream.Stream(cell=tf.keras.layers.Flatten())(net) - streaming of flatten layer - it is not linear low-rank layer

net = tf.keras.layers.Dropout(rate=flags.dropout1)(net) is a dropout regularization.

**Comment by ntyoshi** [comment]
Thank you @rybakov!

How can I make a linear low-rank layer?
I understood [this article on stackExchange](https://stats.stackexchange.com/questions/365179/what-is-low-rank-linear-layer-in-neural-networks),
but I'm not sure how to implement it with Tensorflow.
If I do Flatten() and Dense(), the parameters are too many.

**Comment by rybakov** [comment]
There are two options of doing low rank matrix decomposition:
1. In time and feature domains, well described in [END-TO-END STREAMING KEYWORD SPOTTING](https://arxiv.org/pdf/1812.02802.pdf) it is implemented in [svdf](https://github.com/google-research/google-research/blob/master/kws_streaming/models/svdf.py) and in [ds_tc_resnet](https://github.com/google-research/google-research/blob/master/kws_streaming/models/ds_tc_resnet.py). Here 2D matrix is decomposed into product of two low rank matrices implemented with depthwise convolution (applied in time dimension) and 1x1 convolution (applied in feature dimension).
2. In feature domain only: just represent fully connected layer (e.g 100x100) as two fully connected layers (e.g 100x40 and 40x100).

**Comment by ntyoshi** [comment]
Which model in the paper 1 does the default parameters of [cnn](https://github.com/google-research/google-research/issues/kws_streaming/model/cnn.py) mean?
They are not neither `cnn_trad_fpool3` (CNN-1) nor `cnn_one_fstride4` (CNN-2), right?

**Comment by rybakov** [comment]
Yes, none of the models presented in our [paper](https://arxiv.org/pdf/2005.06720.pdf) have default parameters of [Hello Edge paper](https://arxiv.org/pdf/1711.07128.pdf). In section "4.2. Comparison with baseline" we explicitly say "We improved their accuracy on datasets V1 and V2, as shown on Table 1, by applying SpecAugment with hyper-parameters optimization of both neural net and speech feature extractor parameters".

In Table 1 we labeled models from [Hello Edge paper](https://arxiv.org/pdf/1711.07128.pdf) by tag V1*. Star(*) means that it is baseline accuracy reported by [Hello Edge paper](https://arxiv.org/pdf/1711.07128.pdf). We compared their models with our re-implementations which have different hyperparameters. For example strided CNN ("CNN+strd") parameters are show in [link](https://github.com/google-research/google-research/blob/master/kws_streaming/experiments/kws_experiments_paper_12_labels.md#cnn_stride) are different with hyperparameters of (CNN-2) from [Hello Edge paper](https://arxiv.org/pdf/1711.07128.pdf) implemented in [link](https://github.com/ARM-software/ML-KWS-for-MCU/blob/8151349b110f4d1c194c085fcc5b3535bdf7ce4a/models.py#L643).

If you are interested in reproducing model accuracy from [Hello Edge paper](https://arxiv.org/pdf/1711.07128.pdf), I would suggest to use their [implementations](https://github.com/ARM-software/ML-KWS-for-MCU/blob/8151349b110f4d1c194c085fcc5b3535bdf7ce4a/models.py), so that all hyperparameters will be exactly the same as they are published in [Hello Edge paper](https://arxiv.org/pdf/1711.07128.pdf).

Please feel free to reproduce accuracy of models from [Hello Edge paper](https://arxiv.org/pdf/1711.07128.pdf) with [kws_streaming](https://github.com/google-research/google-research/tree/master/kws_streaming), but as you already noticed you will need to change hyperparameters.

**Comment by ntyoshi** [comment]
Ah I haven't noticed [kws_experiments_paper_12_labels.md](https://github.com/google-research/google-research/blob/master/kws_streaming/experiments/kws_experiments_paper_12_labels.md)!
Thank you for this great environment and your kind responses!
I'm going to reproduce your paper at first.
